fix for #190

petaflot opened the pull request:
also allows encoding of strings and bytestrings (allows encoding any symbol) by converting them to int ; in this case, string representation is shown in human-readable string

petaflot commented:
#190

WhyNotHugo commented on barcode/codex.py:
```diff
@@ -129,7 +129,7 @@ class Code128(Barcode):
     """
 
-    name = "Code 128"
+    ame = "Code 128"
```
?

petaflot replied:
oops :-s

WhyNotHugo commented on barcode/codex.py:
```diff
@@ -271,5 +271,72 @@ def get_fullcode(self):
+    """Initializes a new MSI (Modified Plessey) instance. The checksum is added automatically
+    when building the bars.
+
+    :parameters:
+        code : int or bytes or string
+            Code MSI int without checksum (added automatically).
+            Code MSI bytes without checksum. requires byteorder for int conversion
+            Code MSI string without checksum. requires encoding and byteorder for int conversion
+        writer : barcode.writer Instance
+            The writer to render the barcode (default: SVGWriter).
+        byteorder : string
+            'big' or 'little' ; to convert bytes to int
+        encoding : string
+            if set, convert bytes to string and use this as a label. defaults to utf-8
```
Please use regular sphinx format here so docs render properly.

```suggestion
    :param code: Code MSI int without checksum (added automatically).
            Code MSI bytes without checksum. requires byteorder for int conversion
            Code MSI string without checksum. requires encoding and byteorder for int conversion
    :param writer: The writer to render the barcode (default: SVGWriter).
    :param byteorder: 'big' or 'little' ; to convert bytes to int
    :param encoding : if set, convert bytes to string and use this as a label. defaults to utf-8
```

WhyNotHugo commented on barcode/codex.py:
```diff
@@ -271,5 +271,72 @@ def get_fullcode(self):
+
+    name = "MSI/Modified Plessey"
+
+    def __init__(self, code, writer=None, byteorder=None, encoding=None):
```
Type annotations missing.

petaflot replied:
Can you please help me with this? I don't know how to do it when multiple types are allowed.

I have severe back and joint issues that are made much worse when sitting for extended periods of time, so I will not dig further into an obscure doc to find out.

WhyNotHugo reviewed (commented):
[Wikipedia says](https://en.wikipedia.org/wiki/MSI_Barcode#Character_set_and_binary_lookup) "The MSI bar code represents only digits 0–9; it does not support letters or symbols.".

Is the approach taken for encoding strings a standard one? Do you have a reference for it?

petaflot commented:
it's not standard at all (hence the new defaults). it's a workaround because no 1D barcode that I know of can encode arbitrary bytes (at most a variant of extended ASCII)

the MSI barcode _will_ encode just an int (hence the new default values for "byteorder" and "encoding" : the user _must_ force this workaround, otherwise an exception will be raised when converting individual characters to BCD. I specifically chose to use MSI for this purpose, because since it normally allows only integer encoding the risk of confusion is minimal (user has to explicitly force the conversion : the same approach with ie. Code128 would get very confusing)

does this seem like a reasonable approach? it's a quite inefficient method of encoding data (mostly because of BCD being inefficient in itself) but if the purpose is just to encode short strings it's acceptable. the code could be extended to split longer strings on multiple lines (sort-of like DataBar does) that could be read sequentially.

petaflot commented:
also, from https://www.dynamsoft.com/blog/insights/the-comprehensive-guide-to-1d-and-2d-barcodes/ :
```
MSI Codes are used in inventory management to mark shelves and containers to identify storage locations.

Industry

    Inventory
    Warehousing
```
This basically means that MSI codes are used "internally" by organizations, so it is unlikely that any "consumer" will stumble on such a code (those codes are used in a specific context) ; this is - again - less chance for confusion

petaflot commented:
* added type hints
* removed option to encode strings (it's not standard and other barcode types can do it better)
* default encoding to 'utf-8' again to limit risk of confusion
* user-specifiable label

rmkane commented:
The JsBarcode project includes [multiple implementations of MSI](https://github.com/lindell/JsBarcode/tree/master/src/barcodes/MSI):

- MSI
- MSI10
- MSI1010
- MSI11
- MSI1110

petaflot commented:
> The JsBarcode project includes [multiple implementations of MSI](https://github.com/lindell/JsBarcode/tree/master/src/barcodes/MSI):
> 
>     * MSI
> 
>     * MSI10
> 
>     * MSI1010
> 
>     * MSI11
> 
>     * MSI1110

isn't that javascript?

rmkane commented:
> > The JsBarcode project includes [multiple implementations of MSI](https://github.com/lindell/JsBarcode/tree/master/src/barcodes/MSI):
> > ```
> > * MSI
> > 
> > * MSI10
> > 
> > * MSI1010
> > 
> > * MSI11
> > 
> > * MSI1110
> > ```
> 
> isn't that javascript?

It is, but porting the code over to Python would be an easy task. I was just citing it as an implementation of MSI.
